content/en/docs/plugins/developers_guide.md

```diff
@@ -0,0 +1,1081 @@
+
+![plugin_sdk_go_architecture](/docs/images/plugin_sdk_go_architecture.png)
+
+At a high level, the SDK is on top of three fundamendal package with different levels of abstractions:
+
+1. Package `sdk` is a container for all the basic types, definitions, and helpers that are reused across all the SDK parts. 
+
+2. Package `sdk/symbols` contains prebuilt implementations for all the C symbols that plugins must export to be accepted by the framework. The prebuilt C symbols are divided in many subpackages, so that each of them can be imported individually to opt-in/opt-out each symbol. 
+
+3. Package `sdk/plugins` provide high-level definition and base types for writing extractor plugins and source plugins. This uses `sdk/symbols` internally and takes care of importing all the prebuilt C symbols required by extractor and source plugins respectively. This is the main entrypoint for developers to write plugins in Go.
+
+Two additional packages `ptr` and `cgo` are used internally to simplify and optimize the state management and the usage of C-allocated memory pointers.
+
+For some use cases, developers can consider using the SDK layers selectively. This is meaningful only if developers wish to manually write part of the low-level C details of the framework in their plugins, but still want to use some parts of the SDK. However, this is discouraged if not for advanced use cases only. Developers are encouraged to use the `sdk/plugins` to build Falcosecurity plugins, which is easier to use and will have less frequent breaking changes.
+
+Further details can be found in the documentation of each package: [`sdk`](https://pkg.go.dev/github.com/falcosecurity/plugin-sdk-go/pkg/sdk), [`sdk/symbols`](https://pkg.go.dev/github.com/falcosecurity/plugin-sdk-go/pkg/sdk/symbols), and [`sdk/plugins`](https://pkg.go.dev/github.com/falcosecurity/plugin-sdk-go/pkg/sdk/plugins).
+
+### Getting Started
+The SDK is built on top a set of minimal composable interfaces describing the behavior of plugins and plugin instances. As such, developing plugins is as easy as defining a struct type representing the plugin itself, ensuring that the mandatory interface methods are defined on it, and then registering it to the SDK.
+
+To use the Go SDK, all you need to import are the `sdk` and `sdk/plugins` packages. The first contains all the core types and definitions used across the rest of the SDK packages, whereas the latter contains built-in constructs to develop plugins. The subpackages `sdk/plugins/source` and `sdk/plugins/extractor` contain specialized definitions for source and extractor plugins respectively.
+
+The `dummy` Go plugin, documented in the [next sections](#example-go-plugin-dummy), is a simple example that helps understanding how to start writing Go plugins with this SDK.
+
+#### Defining Extractor Plugins
```

Comment:
When initially reading this section, it made sense--your plugin is either a source or extractor plugin, which each having its own package and interface.

But in the dummy plugin walkthrough, the interfaces (and packages) are more about functionality--a source plugin that extracts fields uses the sdk/plugins/extractor package and interface too.

Can we mention this distinction here as well as in the dummy plugin walkthrough?

Reply:
I agree with you, this creates some ambiguity.

Reply:
I see the changes and I think they help. I'm wondering if we should talk about a "a source plugin with the optional extraction features enabled." though. I can't see a good use case where someone would want to inject custom events but would *not* define fields to extract from those events. The only reason I can think of would be where the event payload is json and all field extraction would be handled by the json plugin. But that's not a great rule writing experience, to force the rule author to keep specifying json pointer paths into the payloads.

I would think the more common case for new source plugins is that they would both inject events and define fields. So why not steer the docs in that direction?

Reply:
That's a good point. Will re-elaborate the paragraphs to apply this change.

Reply:
thanks! the changes help a lot.